metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala

```diff
@@ -1,77 +1,139 @@
+   * 2. test suite's file can be opened
    */
-  private val cachedTestSuites =
-    mutable.Map[FullyQualifiedClassName, TestEntry]()
+  def onChange(docs: TextDocuments, file: AbsolutePath): Unit = {
+    if (isEnabled && index.contains(file)) {
+      if (docs.documents.nonEmpty) {
+        val doc = docs.documents.head
+        index.updateDoc(file, doc)
+      }
+      refreshTestCases(file)
+    }
+  }
 
-  val refreshTestSuites = new BatchedFunction[Unit, Unit](_ =>
-    CancelableFuture(doRefreshTestSuites())
-  )
+  /**
+   * Check if opened file contains test suite and update test cases if yes.
+   */
+  def didOpen(file: AbsolutePath): Future[Unit] =
+    if (isEnabled && index.contains(file)) Future(refreshTestCases(file))
+    else Future.unit
 
   /**
-   * Retrieves cached test suites.
+   * Retrieves all cached test suites. Useful for tests.
    */
-  def findTestSuites(): java.util.List[TestSuiteDiscoveryResult] = {
-    cachedTestSuites.values
-      .groupBy(_.buildTarget)
-      .toSeq
+  def getTestSuites(): java.util.List[BuildTargetUpdate] = {
+    index.suites
       .map { case (buildTarget, entries) =>
-        val grouped = groupTestsByPackage(entries.toList)
-        TestSuiteDiscoveryResult(
-          buildTarget.getDisplayName(),
-          buildTarget.getId().getUri(),
-          grouped.asJava
-        )
+        BuildTargetUpdate(buildTarget, entries.map(_.testClass).toSeq)
       }
+      .toSeq
       .asJava
   }
 
+  /**
+   * Retrieve test cases for a given file. Even an empty list
+   * because it can mean that all testcases were deleted.
+   */
+  def getTestCases(path: AbsolutePath): Seq[BuildTargetUpdate] = {
+    val buildTargetUpdateOpt =
+      for {
+        metadata <- index.getMetadata(path)
+        buildTarget <- metadata.entries.headOption.map(_.buildTarget)
+      } yield {
+        val events = findTestCases(path)
+        BuildTargetUpdate(buildTarget, events)
+      }
+    buildTargetUpdateOpt.toSeq
+  }
+
+  private def refreshTestCases(path: AbsolutePath): Unit = {
+    val buildTargetUpdateOpt = getTestCases(path)
+    buildTargetUpdateOpt.foreach { update =>
+      client.metalsExecuteClientCommand(
+        ClientCommands.UpdateTestExplorer.toExecuteCommandParams(
+          update
+        )
+      )
+    }
+  }
+
+  /**
+   * Searches for test cases for a given path.
+   */
+  private def findTestCases(
+      path: AbsolutePath,
+      symbol: Option[mtags.Symbol] = None
+  ): Seq[AddTestCases] =
+    for {
+      metadata <- index.getMetadata(path).toSeq
+      doc <- metadata.textDocument.toSeq
+      metadataEntry <- metadata.entries
+      testEntry <- index.getTestEntry(
+        metadataEntry.buildTarget,
+        metadataEntry.suiteName
+      )
+      if symbol.forall(_ == testEntry.symbol)
+    } yield {
+      val testClass = testEntry.testClass
+      val testCases = junitTestFinder.findTests(doc, path, testEntry.symbol)
```

Comment:
You should do this onChange and not save semanticdb docs in memory.

Reply:
> not save semanticdb docs in memory.

done, good catch!

> You should do this onChange

I'm doing it, among others, in `onChange` but there are other scenarios where it's useful - for example when the file is currently opened and suite is being discovered.